### ISSUE TITLE
Exporting from a working ScriptModule raises `RuntimeError: nullptr != clone_n INTERNAL ASSERT FAILED` from remove_inplace_ops_for_onnx.cpp

### ISSUE THREAD

**Issue by fxmarty** [pytorch/pytorch#93159]
### 🐛 Describe the bug

TL;DR a ScriptModule model runs fine for inference, but passing it to `torch.onnx.export`  fails with an internal error. @justinchuby @titaiwangms @BowenBao @thiagocrepaldi if you have any suggestion of how I could fix, it would be very helpful! In the meantime I'll try to simplify the model to pin down the issue; stay tuned.

Model: https://huggingface.co/fxmarty/end-to-end-scripted-stable-diffusion/tree/main . If necessary, I can share how the model was generated, but basically: torch.jit.trace a few submodules, and torch.jit.script the top level module of a stable diffusion pipeline largely rewritten to be jit scriptable.

The model can be run as a ScriptModule fine:

```python
import torch

text_input_ids = torch.tensor([[  0, 197,  68, 434,  84, 536, 545, 383, 390, 449, 195,  67,  70, 885,
           1,   1,   1,   1,   1,   1,   1,   1,   1,   1,   1,   1,   1,   1,
           1,   1,   1,   1,   1,   1,   1,   1,   1,   1,   1,   1,   1,   1,
           1,   1,   1,   1,   1,   1,   1,   1,   1,   1,   1,   1,   1,   1,
           1,   1,   1,   1,   1,   1,   1,   1,   1,   1,   1,   1,   1,   1,
           1,   1,   1,   1,   1,   1,   1]])
uncond_text_input_ids = torch.tensor([[0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
         1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
         1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
         1, 1, 1, 1, 1]])
timesteps = torch.tensor([981, 961, 961, 941, 921, 901, 881, 861, 841, 821, 801, 781, 761, 741,
        721, 701, 681, 661, 641, 621, 601, 581, 561, 541, 521, 501, 481, 461,
        441, 421, 401, 381, 361, 341, 321, 301, 281, 261, 241, 221, 201, 181,
        161, 141, 121, 101,  81,  61,  41,  21,   1])

scripted_pipeline = torch.jit.load("scripted_sd.pt")

print(scripted_pipeline.code)
res = scripted_pipeline(
        text_input_ids=text_input_ids,
        uncond_text_input_ids=uncond_text_input_ids,
        timesteps=timesteps,
)
```

The code is as follow:
<details>
  <summary>ScriptModule code</summary>

```
def forward(self,
    text_input_ids: Tensor,
    uncond_text_input_ids: Tensor,
    timesteps: Tensor) -> Tuple[Tensor, bool]:
  sample_size = self.sample_size
  vae_scale_factor = self.vae_scale_factor
  height = torch.mul(sample_size, vae_scale_factor)
  sample_size0 = self.sample_size
  vae_scale_factor0 = self.vae_scale_factor
  width = torch.mul(sample_size0, vae_scale_factor0)
  batch_size = (torch.size(text_input_ids))[0]
  num_images_per_prompt = self.num_images_per_prompt
  prompt_embeds = (self)._encode_prompt(text_input_ids, uncond_text_input_ids, num_images_per_prompt, )
  device = self.device
  unet_in_channels = self.unet_in_channels
  num_images_per_prompt0 = self.num_images_per_prompt
  _0 = torch.mul(batch_size, num_images_per_prompt0)
  _1 = ops.prim.dtype(prompt_embeds)
  latents = (self).prepare_latents(_0, annotate(int, unet_in_channels), annotate(int, height), annotate(int, width), _1, device, )
  _2 = [9223372036854775807, torch.len(timesteps)]
  latents0 = latents
  for i in range(ops.prim.min(_2)):
    t = torch.select(timesteps, 0, i)
    print(i)
    latent_model_input = torch.cat(torch.mul([latents0], 2))
    scheduler = self.scheduler
    latent_model_input0 = (scheduler).scale_model_input(latent_model_input, t, )
    unet = self.unet
    noise_pred = (unet).forward(latent_model_input0, t, prompt_embeds, )
    noise_pred0 = unchecked_cast(Tuple[Tensor], noise_pred)
    noise_pred1 = (noise_pred0)[0]
    noise_pred_uncond, noise_pred_text, = torch.chunk(noise_pred1, 2)
    guidance_scale = self.guidance_scale
    _3 = torch.sub(noise_pred_text, noise_pred_uncond)
    noise_pred2 = torch.add(noise_pred_uncond, torch.mul(_3, guidance_scale))
    scheduler0 = self.scheduler
    _4 = (scheduler0).step(noise_pred2, annotate(int, t), latents0, )
    latents0 = (_4)[0]
  image = (self).decode_latents(latents0, )
  return (image, True)
```

where `_encode_prompt.code` is:

```
def _encode_prompt(self,
    text_input_ids: Tensor,
    uncond_text_input_ids: Tensor,
    num_images_per_prompt: int) -> Tensor:
  batch_size = (torch.size(text_input_ids))[0]
  text_encoder = self.text_encoder
  prompt_embeds = (text_encoder).forward(text_input_ids, )
  prompt_embeds0 = (prompt_embeds)[0]
  bs_embed, seq_len, _0, = torch.size(prompt_embeds0)
  prompt_embeds1 = torch.repeat(prompt_embeds0, [1, num_images_per_prompt, 1])
  _1 = torch.mul(bs_embed, num_images_per_prompt)
  prompt_embeds2 = torch.view(prompt_embeds1, [_1, seq_len, -1])
  text_encoder0 = self.text_encoder
  negative_prompt_embeds = (text_encoder0).forward(uncond_text_input_ids, )
  negative_prompt_embeds0 = (negative_prompt_embeds)[0]
  seq_len0 = (torch.size(negative_prompt_embeds0))[1]
  negative_prompt_embeds1 = torch.repeat(negative_prompt_embeds0, [1, num_images_per_prompt, 1])
  _2 = torch.mul(batch_size, num_images_per_prompt)
  negative_prompt_embeds2 = torch.view(negative_prompt_embeds1, [_2, seq_len0, -1])
  _3 = [negative_prompt_embeds2, prompt_embeds2]
  return torch.cat(_3)
```

where `decode_latents.code` is:

```
def decode_latents(self,
    latents: Tensor) -> Tensor:
  latents2 = torch.mul(latents, 5.4899807850672522)
  vae_decoder = self.vae_decoder
  image = (vae_decoder).forward(latents2, )
  image0 = (image)[0]
  image1 = torch.clamp(torch.add(torch.div(image0, 2), 0.5), 0, 1)
  return torch.permute(image1, [0, 2, 3, 1])
```

where `scheduler.step.code` is:
```
def step(self,
    model_output: Tensor,
    timestep: int,
    sample: Tensor) -> Tuple[Tensor]:
  counter = self.counter
  prk_timesteps = self.prk_timesteps
  _0 = torch.lt(counter, torch.len(prk_timesteps))
  if _0:
    skip_prk_steps = self.skip_prk_steps
    _1 = torch.__not__(skip_prk_steps)
  else:
    _1 = False
  if _1:
    _3 = (self).step_prk(model_output, timestep, sample, )
    _2 = _3
  else:
    _4 = (self).step_plms(model_output, timestep, sample, )
    _2 = _4
  return _2
```

And I could go on with submodules...

----
</details>

However the ONNX export fails:
```python
torch.onnx.export(
    scripted_pipeline,
    args=(text_input_ids, uncond_text_input_ids, timesteps),
    f="stable_diffusion_pipeline.onnx",
    input_names=["text_input_ids", "uncond_text_input_ids", "timesteps"],
)
```

with:

```
  File "/home/fxmarty/hf_internship/optimum/research/stable_diffusion_end_to_end_onnx/onnx_export.py", line 42, in <module>
    torch.onnx.export(
  File "/home/fxmarty/anaconda3/envs/hf-inf/lib/python3.9/site-packages/torch/onnx/utils.py", line 504, in export
    _export(
  File "/home/fxmarty/anaconda3/envs/hf-inf/lib/python3.9/site-packages/torch/onnx/utils.py", line 1529, in _export
    graph, params_dict, torch_out = _model_to_graph(
  File "/home/fxmarty/anaconda3/envs/hf-inf/lib/python3.9/site-packages/torch/onnx/utils.py", line 1115, in _model_to_graph
    graph = _optimize_graph(
  File "/home/fxmarty/anaconda3/envs/hf-inf/lib/python3.9/site-packages/torch/onnx/utils.py", line 617, in _optimize_graph
    _C._jit_pass_onnx_remove_inplace_ops_for_onnx(graph, module)
RuntimeError: nullptr != clone_n INTERNAL ASSERT FAILED at "../torch/csrc/jit/passes/onnx/remove_inplace_ops_for_onnx.cpp":197, please report a bug to PyTorch. 
```

coming from https://github.com/pytorch/pytorch/blob/62aa4e096b8414ad22796d0dbffa12988a9f3793/torch/csrc/jit/passes/onnx/remove_inplace_ops_for_onnx.cpp#L212

### Versions

PyTorch version: 1.13.1+cu117
Is debug build: False
CUDA used to build PyTorch: 11.7
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.1 LTS (x86_64)
GCC version: (Ubuntu 11.3.0-1ubuntu1~22.04) 11.3.0
Clang version: 14.0.0-1ubuntu1
CMake version: version 3.25.0
Libc version: glibc-2.35

Python version: 3.9.12 (main, Apr  5 2022, 06:56:58)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-5.15.0-58-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 11.7.99
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 3060 Laptop GPU
Nvidia driver version: 515.86.01
cuDNN version: Probably one of the following:
/usr/local/cuda-11.7/targets/x86_64-linux/lib/libcudnn.so.8.7.0
/usr/local/cuda-11.7/targets/x86_64-linux/lib/libcudnn_adv_infer.so.8.7.0
/usr/local/cuda-11.7/targets/x86_64-linux/lib/libcudnn_adv_train.so.8.7.0
/usr/local/cuda-11.7/targets/x86_64-linux/lib/libcudnn_cnn_infer.so.8.7.0
/usr/local/cuda-11.7/targets/x86_64-linux/lib/libcudnn_cnn_train.so.8.7.0
/usr/local/cuda-11.7/targets/x86_64-linux/lib/libcudnn_ops_infer.so.8.7.0
/usr/local/cuda-11.7/targets/x86_64-linux/lib/libcudnn_ops_train.so.8.7.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.22.4
[pip3] torch==1.13.1
[pip3] torch-model-archiver==0.6.1
[pip3] torch-workflow-archiver==0.2.5
[pip3] torchinfo==1.7.0
[pip3] torchserve==0.6.1
[pip3] torchtriton==2.0.0+0d7e753227
[pip3] torchvision==0.14.1
[conda] cudatoolkit               11.3.1               h2bc3f7f_2    anaconda
[conda] numpy                     1.22.4                   pypi_0    pypi
[conda] torch                     1.13.1                   pypi_0    pypi
[conda] torch-model-archiver      0.6.1                    pypi_0    pypi
[conda] torch-workflow-archiver   0.2.5                    pypi_0    pypi
[conda] torchinfo                 1.7.0                    pypi_0    pypi
[conda] torchserve                0.6.1                    pypi_0    pypi
[conda] torchtriton               2.0.0+0d7e753227          pypi_0    pypi
[conda] torchvision               0.14.1                   pypi_0    pypi

**Comment by BowenBao** [comment]
Hello @fxmarty, thank you for reporting, this is an interesting model! I think the issue may be coming from attribute `ets` of the `scheduler`. 
```python
    def step_plms(
        self,
        model_output: torch.FloatTensor,
        timestep: int,
        sample: torch.FloatTensor,
        return_dict: bool = True,
    ) -> Union[SchedulerOutput, Tuple]:
    ...
        if self.counter != 1:
            self.ets = self.ets[-3:]
            self.ets.append(model_output)
```
There are limitation handling assignment of list attributes along with slicing within controlflow. I wonder if `ets` can be modified to a tensor? As long as `model_output` have the same shape, the operations should still be compatible.

**Comment by fxmarty** [comment]
@BowenBao Thanks a lot for you quick feedback! I'll have a look shortly.

**Comment by fxmarty** [comment]
@BowenBao The issue indeed comes from the line `latents = self.scheduler.step(noise_pred, t, latents)[0]`, commenting it out the export goes fine.

It appears the issue was rather with type hinting `Optional[torch.Tensor]` used to have torch.jit.script pass. `self.cur_sample` was hinted as `Optional[torch.Tensor]`, thus requiring `sample` to be hinted `Optional[torch.Tensor]` as well. However, since it is used in a pytorch ops later on, I used `assert sample is not None` for torch.jit.script to pass. It appears the `self.cur_sample = None` made the export fail.

As a solution, I use `torch.tensor(-1)` instead of None and type everything as `torch.Tensor`.

I also use a buffer for `self.ets`, it's indeed cleaner. Just hope that `torch.roll` is not too slow when exported.

Relevant code sample: https://github.com/huggingface/diffusers/blob/a6610db7a8ea0c8b04ec915b8ff3cac670606827/src/diffusers/schedulers/scheduling_pndm.py#L314-L336
